server/src/main/java/org/opensearch/rest/action/cat/RestIndicesAction.java

```diff
@@ -205,6 +218,91 @@ public void onFailure(final Exception e) {
         };
     }
 
+    public RestChannelConsumer doPaginatedCatRequest(
+        final RestRequest request,
+        final NodeClient client,
+        final TimeValue clusterManagerNodeTimeout,
+        final String[] indices
+    ) {
+        final boolean local = request.paramAsBoolean("local", false);
+        final boolean includeUnloadedSegments = request.paramAsBoolean("include_unloaded_segments", false);
+        final String requestedToken = request.param("next_token");
+        IndexBasedPaginationStrategy.validateRequestedRequest(requestedToken);
+        final int pageSize = Integer.parseInt(request.param("max_page_size", DEFAULT_CAT_INDICES_PAGE_SIZE_STRING));
+        final boolean latestIndicesFirst = request.paramAsBoolean("latest_indices_first", false);
+
+        return channel -> {
```

Comment:
I think we should avoid this duplication of code here and doCatRequest. 

We can take additional parameter to the doCatRequest method :- PageFilterSupplier (ClusterState)
The implementation of this can be NoOp for non-paginated query while it can take Request+pageToken + ClusterState to return list of indices.